File: app/src/androidTest/java/com/example/java_unittest_mitch/persistence/NoteDaoTest.java
package com.example.java_unittest_mitch.persistence;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;

import android.database.sqlite.SQLiteConstraintException;
import android.text.TextUtils;
import android.util.Log;

import androidx.arch.core.executor.testing.InstantTaskExecutorRule;
import androidx.lifecycle.LiveData;

import com.example.java_unittest_mitch.models.Note;
import com.example.java_unittest_mitch.util.LiveDataTestUtil;
import com.example.java_unittest_mitch.util.TestUtil;

import org.junit.Assert;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.mockito.internal.matchers.LessOrEqual;
import org.mockito.internal.matchers.Not;

import java.util.List;

public class NoteDaoTest extends NoteDatabaseTest
{

    @Rule
    public InstantTaskExecutorRule rule=new InstantTaskExecutorRule();
    //this doesn't work in junit5
    /*
        Insert,Read,Delete
     */
    @Test
    public void testInsertReadAndDelete() throws Exception {
        //insert
        Note note1=new Note(TestUtil.TEST_NOTE_1);

        getNoteDao().insertNote(note1).blockingGet();

        //read
        LiveDataTestUtil<List<Note>> liveData=new LiveDataTestUtil<>();
        List<Note> insertedNotes=liveData.getValue(getNoteDao().getNotes());
        assertNotNull(insertedNotes);
        note1.setId(insertedNotes.get(0).getId());
        assertEquals(note1.getTitle(),insertedNotes.get(0).getTitle());
        assertEquals(note1.getContent(),insertedNotes.get(0).getContent());
        assertEquals(note1.getTimeStamp(),insertedNotes.get(0).getTimeStamp());
        assertEquals(note1,insertedNotes.get(0));
        //delete
        getNoteDao().deleteNote(note1).blockingGet();
        insertedNotes=liveData.getValue(getNoteDao().getNotes());
        assertEquals(0,insertedNotes.size());

    }
    /*
        Insert,Read,Update,Read,Delete
     */
    @Test
    public void testInsertReadUpdateReadDelete() throws Exception {
        //insert
        Note note=new Note(TestUtil.TEST_NOTE_2);
        getNoteDao().insertNote(note).blockingGet();
        //read
        LiveDataTestUtil<List<Note>> liveDataTestUtil=new LiveDataTestUtil<>();
        List<Note> notes=liveDataTestUtil.getValue(getNoteDao().getNotes());
        assertNotNull(notes);
        //update
        note.setTitle("title1");
        note.setContent("content1");
        note.setTimeStamp("today");
        note.setId(notes.get(0).getId());

        getNoteDao().updateNote(note).blockingGet();
        //read
        notes=liveDataTestUtil.getValue(getNoteDao().getNotes());
        Log.i("mag2851",notes.get(0).getTitle()+"\n"+note.getTitle());

        assertEquals(note.getTitle(),notes.get(0).getTitle());
        assertEquals(note.getContent(),notes.get(0).getContent());
        assertEquals(note.getTimeStamp(),notes.get(0).getTimeStamp());

        //delete
        getNoteDao().deleteNote(note).blockingGet();
        notes=liveDataTestUtil.getValue(getNoteDao().getNotes());
        assertEquals(0,notes.size());

    }


    /*
        Insert note with null title => throws excpetion
     */

    @Test(expected = SQLiteConstraintException.class)
    public void insertNotWithNullTitle() throws Exception {
        Note note=new Note(1,null,"content","time");
        getNoteDao().insertNote(note).blockingGet();



    }

    /*
        Insert,Read,Update with null title => throws exception
     */
    @Test(expected = SQLiteConstraintException.class)
    public void insertReadAndUpdateWithNull() throws Exception {
        //insert
        Note note=new Note(TestUtil.TEST_NOTE_1);
        getNoteDao().insertNote(note).blockingGet();
        LiveDataTestUtil<List<Note>> liveDataTestUtil=new LiveDataTestUtil<>();
        List<Note> notes=liveDataTestUtil.getValue(getNoteDao().getNotes());

        //read
        assertNotNull(notes);
        //update title with null
        note.setTitle(null);
        note.setId(notes.get(0).getId());
        getNoteDao().updateNote(note).blockingGet();


    }

}

File: app/src/androidTest/java/com/example/java_unittest_mitch/persistence/NoteDaoTest.kt
package com.example.java_unittest_mitch.persistence

import android.database.sqlite.SQLiteConstraintException
import android.util.Log
import androidx.arch.core.executor.testing.InstantTaskExecutorRule
import com.example.java_unittest_mitch.TIME1
import com.example.java_unittest_mitch.models.Note
import com.example.java_unittest_mitch.testNote
import com.example.java_unittest_mitch.util.LiveDataTestUtil
import org.junit.Assert
import org.junit.Rule
import org.junit.Test

class NoteDaoTest:NoteDatabaseTest()
{
    @get:Rule
    val rule=InstantTaskExecutorRule()

    /*Insert,Read,Delete*/
    @Test
    fun insert_Read_delete_A_NOTE()
    {
        val note= testNote
        //---------------
        //insert
        getNoteDao().insertNote(note).blockingGet()
        //----------
        //read
        val liveUtil=LiveDataTestUtil<List<Note>>()
        var notes=liveUtil.getValue(getNoteDao().getNotes())
        Assert.assertNotNull(notes)//check inserted note

        Log.i("mag2851-noteSize-1",notes?.size.toString())
        Assert.assertEquals(notes?.get(0)?.title,note.title)
        Assert.assertEquals(notes?.get(0)?.content,note.content)
        Assert.assertEquals(notes?.get(0)?.timeStamp,note.timeStamp)

        note.id= notes?.get(0)?.id!!

        Assert.assertEquals(note,notes.get(0))
        //------------------------------------------

        Log.i("mag2851-noteSize-2",notes?.size.toString())
        //delete
        getNoteDao().deleteNote(note).blockingGet()
        notes=liveUtil.getValue(getNoteDao().getNotes())
        Log.i("mag2851-noteSize-3",notes?.size.toString())
        Assert.assertEquals(notes?.size,0)
        //-----------------------------------------------
    }
    /*-----------------------------------------------*/
    /*Insert,Read,Update,Read,Delete*/
    @Test
    fun insert_read_update_read_delete_A_Note()
    {
        val noteModel= testNote
        //Insert
        getNoteDao().insertNote(noteModel).blockingGet()
        val liveDataTestUtil=LiveDataTestUtil<List<Note>>()
        var notesData=liveDataTestUtil.getValue(getNoteDao().notes)
        Assert.assertEquals(notesData?.size,1)
        //Update----------------------------------------------
        noteModel.title="majid2851"
        noteModel.content="this is my new note"
        noteModel.timeStamp="12-1999"
        noteModel.id=notesData?.get(0)?.id!!
        getNoteDao().updateNote(noteModel).blockingGet()
        notesData=liveDataTestUtil.getValue(getNoteDao().notes)
        Assert.assertEquals(noteModel,notesData?.get(0))
        //Delete
        getNoteDao().deleteNote(noteModel).blockingGet()
        notesData=liveDataTestUtil.getValue(getNoteDao().notes)
        Assert.assertEquals(notesData?.size,0)

    }
    /*----------------------------------------------*/
    /*Insert with null Title,throw Excepti on*/
    @Test(expected = SQLiteConstraintException::class)
    fun insertWithNullTitle()
    {
        val noteModel=Note()
        noteModel.timeStamp= TIME1
        noteModel.content="this note has a null value"
        getNoteDao().insertNote(noteModel).blockingGet()
    }
    /*----------------------------------------------*/
    /*Insert,Update with null Title,throw Exception*/
    @Test(expected = SQLiteConstraintException::class)
    fun insert_update_withNUll_throwExcption()
    {
        var note= Note()
        getNoteDao().insertNote(note).blockingGet()
        val liveDataTestUtil=LiveDataTestUtil<List<Note>>()
        val notes=liveDataTestUtil.getValue(getNoteDao().notes)
        note.id= notes?.get(0)?.id!!
        note.content="content"
        note.timeStamp="05-2022"
        getNoteDao().updateNote(note).blockingGet()


    }

    /*----------------------------------------------*/

}